File: composeApp/src/commonMain/kotlin/com/daddoodev/yetimatch/services/QuizService.kt
package com.daddoodev.yetimatch.services

import com.daddoodev.yetimatch.models.Quiz
import com.daddoodev.yetimatch.models.QuizManifest
import com.daddoodev.yetimatch.models.QuizResult
import kotlinx.serialization.json.Json

class QuizService {
    private val json = Json {
        ignoreUnknownKeys = true
        prettyPrint = true
    }

    fun loadManifest(jsonString: String): QuizManifest {
        return json.decodeFromString<QuizManifest>(jsonString)
    }

    /**
     * Load a quiz from JSON string
     */
    fun loadQuiz(jsonString: String): Quiz {
        return json.decodeFromString<Quiz>(jsonString)
    }
    
    /**
     * Calculate quiz result based on answers
     * This uses the simple direct mapping approach we discussed
     */
    fun calculateResult(quiz: Quiz, selectedAnswers: Map<String, String>): QuizResult {
        // Count votes for each character
        val characterVotes = mutableMapOf<String, Int>()
        
        selectedAnswers.forEach { (questionId, answerId) ->
            // Find the question
            val question = quiz.questions.find { it.id == questionId }
            
            // Find the answer
            val answer = question?.answers?.find { it.id == answerId }
            
            // Increment vote count for this character
            answer?.let {
                characterVotes[it.characterId] = (characterVotes[it.characterId] ?: 0) + 1
            }
        }
        
        // Find the character with the most votes
        val winningCharacterId = characterVotes.maxByOrNull { it.value }?.key
            ?: quiz.results.first().characterId  // Fallback to first result
        
        // Return the corresponding result
        return quiz.results.find { it.characterId == winningCharacterId }
            ?: quiz.results.first()  // Fallback to first result
    }
    
    /**
     * Get all available quizzes
     * For now returns a list, but you could load multiple quiz files
     */
    fun getAvailableQuizzes(): List<Quiz> {
        // In a real app, you'd load from multiple JSON files in assets
        // For now, return empty list - you'll populate this when loading from assets
        return emptyList()
    }
}

File: lib/services/quiz_service.dart
import 'dart:convert';

import '../models/manifest.dart';
import '../models/quiz.dart';

class QuizService {
  QuizManifest loadManifestFromJson(String jsonString) {
    final map = jsonDecode(jsonString) as Map<String, dynamic>;
    return QuizManifest.fromJson(map);
  }

  Quiz loadQuizFromJson(String jsonString) {
    final map = jsonDecode(jsonString) as Map<String, dynamic>;
    return Quiz.fromJson(map);
  }

  QuizResult calculateResult(Quiz quiz, Map<String, String> selectedAnswers) {
    final characterVotes = <String, int>{};
    for (final e in selectedAnswers.entries) {
      final qList = quiz.questions.where((q) => q.id == e.key).toList();
      final question = qList.isEmpty ? null : qList.first;
      final aList = question?.answers.where((a) => a.id == e.value).toList();
      final answer = (aList == null || aList.isEmpty) ? null : aList.first;
      if (answer != null) {
        characterVotes[answer.characterId] =
            (characterVotes[answer.characterId] ?? 0) + 1;
      }
    }
    if (characterVotes.isEmpty) {
      return quiz.results.first;
    }
    final winningId = characterVotes.entries
        .reduce((a, b) => a.value >= b.value ? a : b)
        .key;
    return quiz.results.firstWhere(
      (r) => r.characterId == winningId,
      orElse: () => quiz.results.first,
    );
  }
}
